enhancements/kube-apiserver/generic-delete-protection.md

```diff
@@ -0,0 +1,233 @@
+replaces:
+superseded-by:
+---
+
+# Generic Delete Protection
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions 
+
+1. At this moment, adoption is racy because deployment.apps/name is insufficient to distinguish based on UID, but to allow
+   freedom of creation ordering, the CriticalService cannot include a UID.
+   We may be able to create a status field that tracks a UID, but the intent isn't fully clear.
+   If adoption doesn't work, we can end up in a state where ManifestWork cannot be finalized because the CriticalService
+   isn't actually eligible for deletion.
+   The simplest solution seem to be to state an intent of "best-effort-deletion" or "orphan-resources".
+   This could be added as a spec field which could be set post-deletion, pre-finalization.
+
+## Summary
+
+By running the platform on the cluster, there are complications that arise regarding deletion of certain critical services.
+These services include webhook admission plugins and finalizers.
+This design is about trying to find a way to prevent permanently "stuck" resources without trying to coordinate resource deletion
+in clients.
+
+## Motivation
+
+If pods for admission-webhook-a, which protect resource-a, are deleted, then it becomes impossible to create, update, or
+delete any resource-a until the webhookconfiguration is deleted or the pods are restored.
+If pods for finalizer-a, which protect resource-a, are deleted, then it becomes impossible to delete any resource-a until
+the pods are restored.
+Resolving resource deletion ordering in clients is impractical.
+
+### Goals
+
+1. Eliminate the need for ordered client deletion.
+
+### Non-Goals
+
+1. Prevent permanently stuck resources in cases where the resource that is stuck and the the pod providing the critical service
+   are in the same namespace.  Don't do that. 
+
+## Proposal
+
+We will create a validating admission webhook that intercepts DELETEs of namespaces, criticalservices, and any resource listed as a .spec.provider.
+Realistically, we will hard code deployments.apps and have that as the only valid resource in .spec.provider to start.
+When a DELETE arrives for this resource we will..
+
+1. Check all CriticalServices to see if this particular instance is protected, if not ALLOW the delete.
+2. Check the .spec.criteria for each matching CriticalService.
+   If all .spec.criteria are satisfied, ALLOW the delete.
+   If not all .spec.criteria are satisfied, do something sane in spec and DENY the delete.
+3. As a special case, a CriticalService cannot be deleted until its .spec.provider is no longer present.
+
+```go
+package api
+
+type CriticalService struct{
```

Comment:
the "service" term is overloaded. CriticalResource? CriticalComponent?

In case of service, I would also expect multiple providers. There could be a deployment, a service and more which contribute to that service.